spec.md

```diff
@@ -179,6 +218,43 @@ physically present in the final circuit. Refer to the description of the
 instance statement for details on how to instantiate a module
 ([@sec:instances]).
 
+### Public Modules
+
+A public module is a module that is marked with the `public`{.firrtl} keyword.
+This means that this is a module that is usable external to this circuit.  A
+public module thereby has a defined ABI---knowing the FIRRTL representation is
+sufficient to know its compiled representation.
+
+The example below shows a public module `Foo`:
+
+``` firrtl
+public module Foo:
+  input a: UInt<1>
+  output b: UInt<1>
+
+  connect b, a
+```
+
+A public module has a number of restrictions:
+
+1. A public module may have no ports of uninferred width.
+1. A public module may have no ports of abstract reset type.
+1. A public module may have no ports of input probe type.
+1. A `RWProbe`{.firrtl} may not be used to access a public module's ports.
```

Comment:
at some point we're going to need to relax this RWProbe thing...

Reply:
In short, this is a rule because of the net propagation problems that violate basic FIRRTL semantics and cause problems in practice that can be a pain to track down (and are not consistently simulated).

So IMO it's not about relaxing it's about not breaking things that need to work.  What motivates this comment / what's the ask?

Since you brought it up :wink:.... none of this is new but perhaps needs reviving:

As long as we don't emit ports as variables (logic), and as nets (as we do now), I think we want that restriction.  I'm not sure how reasonable or not it'd be to move forward with such a change.  Do you have a sense of that?
I think this was discussed long ago and maybe even attempted but I don't know the details.

One idea: require as part of the extmodule/module signature, and ABI, that it may be forced and undefined behavior to have it wrong re:forceability on either side of the boundary.  This could in practice be "emit as variable", but selectively not everywhere (in trade for requiring correct declarations or bugs), basically encode the forceability on the type itself (not the rwprobe, the target port).  This doesn't really feel like a win, but maybe others have other opinions.

Or: we let the net propagation problems be users problem/implementation (/simulator-)defined behavior, hopefully giving them a way to mitigate that isn't manually inserting buffer wires and marking don't touch in order to use this feature or in defense of possibility it's used in a module they don't own (this happens in practice in designs "in the wild" due to adhoc use of force XMR's).  This has benefit of not getting in the way of users who don't want this restriction, but is my least favorite as I'd rather make RWProbe a useful robust concept that acts as folks might expect it to.

WDYT? Is there a way to get what drives this ask a different way?

Basically it's for those reasons it's a requirement, no disagreement that it'd be better to not have such a rule if there's a better place we can take things :+1: .

Reply:
Anyway this probably should be discussed elsewhere, but don't mean to close this down -- please LMK here or elsewhere if have any further thoughts/ideas/suggestions.

Reply:
Ah, I misread this as "there may not be any RWProbe type ports". Which is not in fact what this says, and now I understand your objection.